store/postgres/src/deployment_store.rs

```diff
@@ -1025,32 +1038,57 @@ impl DeploymentStore {
         &self,
         logger: &Logger,
         site: Arc<Site>,
-        graft_base: Option<(Arc<Site>, EthereumBlockPointer)>,
+        graft_src: Option<(Arc<Layout>, EthereumBlockPointer)>,
     ) -> Result<(), StoreError> {
-        let conn = self.get_conn()?;
+        let dst = self.find_layout(site)?;
 
         // Do any cleanup to bring the subgraph into a known good state
-        conn.transaction(|| {
-            let layout = self.layout(&conn, site.clone())?;
-            if let Some((base, block)) = graft_base {
-                let base_layout = self.layout(&conn, base.clone())?;
+        if let Some((src, block)) = graft_src {
+            let start = Instant::now();
+
+            info!(
+                logger,
+                "Initializing graft by copying data from {} to {}",
+                src.catalog.namespace,
+                dst.catalog.namespace
+            );
+
+            // 1. Copy subgraph data
+            // We allow both not copying tables at all from the source, as well
+            // as adding new tables in `self`; we only need to check that tables
+            // that actually need to be copied from the source are compatible
+            // with the corresponding tables in `self`
+            let copy_conn = crate::copy::Connection::new(self.conn.clone());
+            copy_conn.copy_data(logger, src.clone(), dst.clone(), block.clone())?;
+
+            let conn = self.get_conn()?;
+            conn.transaction(|| -> Result<(), StoreError> {
+                // 2. Copy dynamic data sources and adjust their ID
+                let count = dynds::copy(&conn, &src.site, &dst.site)?;
+                info!(logger, "Copied {} dynamic data sources", count;
+                      "time_ms" => start.elapsed().as_millis());
+
+                // 3. Rewind the subgraph. `revert_block` gets rid of everything
+                // including the block passed to it. We want to preserve `block`
+                // and therefore revert `block+1`
                 let start = Instant::now();
-                layout.copy_from(
-                    logger,
-                    &conn,
-                    &site.deployment,
-                    &base_layout,
-                    &base.deployment,
-                    block.clone(),
-                )?;
+                let block_to_revert: BlockNumber = (block.number + 1)
+                    .try_into()
+                    .expect("block numbers fit into an i32");
+                dst.revert_block(&conn, &dst.site.deployment, block_to_revert)?;
+                Layout::revert_metadata(&conn, &dst.site.deployment, block_to_revert)?;
+                info!(logger, "Rewound subgraph to block {}", block.number;
+                      "time_ms" => start.elapsed().as_millis());
```

Comment:
It'd be good to see more justification for this step in the comment.

Reply:
I've actually been meaning to get rid of this and simply forgot. It was left over from when we copied everything and then got rid of the excess because copying the complex metadata structure up to just some block was too cumbersome. I've removed that in favor of copying just what we need.

Reply:
As it turns out, reverting the data is important since it needs to unclamp versions that are deleted in the future. In theory, `CopyEntityBatchQuery` could do that during copying, but the SQL to do that becomes pretty unwieldy. Maybe as an optimization in the future.

Reply:
Ooh tricky